app/waf/src/repositories/common.ts

```diff
@@ -15,11 +24,80 @@ export class CommonRepository<
     return super.create(data, options);
   }
 
-  updateById(id: ID, data: DataObject<T>, options?: Options): Promise<void> {
-    Object.assign(data, {
-      updatedAt: new Date().toISOString(),
-    });
+  find(filter?: Filter<T>, options?: Options): Promise<T[]> {
+    if (options !== undefined && options['tenantId']) {
+      let tenantFilter: Filter = {where: {tenantId: options['tenantId']}};
+      return super.find(tenantFilter as (Filter<T>));
+    } else {
+      return super.find(filter, options);
+    }
+  }
+
+  async findById(id: ID, filter?: Filter<T>, options?: Options): Promise<T> {
+    if (options !== undefined && options['tenantId']) {
+      let tenantFilter: Filter = {
```

Comment:
LB4 Filter where clause seems to support building new where clause from existing one. Did you have a try?

Reply:
I write test code like below:
```
  @get(prefix + '/monitors/{monitorId}', {
    responses: {
      responses: {
        '200': Schema.response(Monitor, 'Successfully retrieve Pool resource'),
        '404': Schema.notFound('Can not find Pool resource'),
      },
    },
  })
  async findById(
    @param(Schema.pathParameter('monitorId', 'Monitor resource ID'))
    id: string,
  ): Promise<Response> {
    let tenantId = "xxxxxxxxxxx";
    console.log("*******************");
    console.log("a fake tenantId " + tenantId);
    console.log("*******************");
    const data = await this.monitorRepository.findById(
      id,
      { where: { tenantId: tenantId } },
    );
    return new Response(Monitor, data);
  }
```

and call this API
```
http://127.0.0.1:3000/adcaas/v1/monitors/fa1f961a-6681-4319-94d8-6478472bc447
```

I got a response as below:
```
{
    "monitor": {
        "id": "fa1f961a-6681-4319-94d8-6478472bc447",
        "name": "t3",
        "description": "My description",
        "tenantId": "fdac59f5b20046829ea58720702a74af",
        "createdAt": "2019-04-28T02:39:28.477Z",
        "updatedAt": null,
        "interval": 5,
        "timeout": 16,
        "targetAddress": "192.168.10.123",
        "targetPort": 8080,
        "monitorType": "tcp"
    }
}
```

the loopback4 log is as below:
```
Server is running at http://127.0.0.1:3000
Try http://127.0.0.1:3000/adcaas/v1/ping
2019-04-30 12:06:22,177 INFO [api.call] Request: {"uuid":"5077a800-6afd-11e9-96e3-af1eed1f6296","method":"GET","headers":{"x-auth-token":"gAAAAABcx8a65YCYLc0LXk6z3FjUDN0qpY2zwIoGLQ4QGohK8XJ6LoSr7DIFzUgRIpLTl7w5fgfrDk-PH-SB14kxqdfNuq1L6KapE7S4KRd_vNC0q841xbBgVCt4jdyIuroA7PbR0AqGBOio6OqHahI2YaPjZKMniDRU8Nwcj1O7iyKAj3wzB84","tenant-id":"fdac59f5b20046829ea58720702a74af","content-type":"application/json","user-agent":"PostmanRuntime/7.11.0","accept":"*/*","cache-control":"no-cache","postman-token":"02069883-953c-4771-88ca-e5d0502d24fd","host":"127.0.0.1:3000","accept-encoding":"gzip, deflate","connection":"keep-alive"},"path":"/adcaas/v1/monitors/fa1f961a-6681-4319-94d8-6478472bc447","params":{}}
2019-04-30 12:06:22,179 DEBUG [api.call] start to authenticate user
2019-04-30 12:06:23,503 DEBUG [api.call] Authenticated OK
2019-04-30 12:06:23,503 DEBUG [api.call] {"region":"RegionOne","interface":"internal","version":"v3","issuedAt":"2019-04-30T03:53:30.000Z","expiredAt":"2019-04-30T04:53:30.000Z","token":"gAAAAABcx8a65YCYLc0LXk6z3FjUDN0qpY2zwIoGLQ4QGohK8XJ6LoSr7DIFzUgRIpLTl7w5fgfrDk-PH-SB14kxqdfNuq1L6KapE7S4KRd_vNC0q841xbBgVCt4jdyIuroA7PbR0AqGBOio6OqHahI2YaPjZKMniDRU8Nwcj1O7iyKAj3wzB84","userId":"934d8da3f75d4bf3b07375a8701f29ca","catalog":[{"endpoints":[{"url":"http://10.145.65.10/placement","interface":"internal","region":"RegionOne","region_id":"RegionOne","id":"0f75bfa5baf64ddca7a30dd4bd0f7af8"},{"url":"http://10.145.65.10/placement","interface":"public","region":"RegionOne","region_id":"RegionOne","id":"34a83f5d7e094965988d3c36311980f6"},{"url":"http://10.145.65.10/placement","interface":"admin","region":"RegionOne","region_id":"RegionOne","id":"ab5d31ddc96c47fbba1d97c836812c6e"}],"type":"placement","id":"1f15566d94c742938bc3d8ef8cdc74e5","name":"placement"},{"endpoints":[{"url":"http://10.145.65.10:8774/v2.1","interface":"internal","region":"RegionOne","region_id":"RegionOne","id":"6b663a069cab4d8d8626a87e5a9cfceb"},{"url":"http://10.145.65.10:8774/v2.1","interface":"admin","region":"RegionOne","region_id":"RegionOne","id":"963eca6fa7854da299ab443a6fc5e1cd"},{"url":"http://10.145.65.10:8774/v2.1","interface":"public","region":"RegionOne","region_id":"RegionOne","id":"99ff2ca2b0b74bfb96762d578bd6d75b"}],"type":"compute","id":"38c02cac0cc842de9cbf2ca97fd41131","name":"nova"},{"endpoints":[{"url":"http://10.145.65.10:9696/","interface":"admin","region":"RegionOne","region_id":"RegionOne","id":"5cfb9cb93cac42d49b62268184d20fc1"},{"url":"http://10.145.65.10:9696/","interface":"internal","region":"RegionOne","region_id":"RegionOne","id":"9297437c4c93492bb9edc7df3a7c8436"},{"url":"http://10.145.65.10:9696/","interface":"public","region":"RegionOne","region_id":"RegionOne","id":"ca1791b2656440819be7bb8c0857fd80"}],"type":"network","id":"8a52dd11d0b14fc1bac7444b7fb6435f","name":"neutron"},{"endpoints":[{"url":"http://10.145.65.10/identity","interface":"public","region":"RegionOne","region_id":"RegionOne","id":"9bf7680f124747088dff9447d7144ee2"},{"url":"http://10.145.65.10/identity","interface":"internal","region":"RegionOne","region_id":"RegionOne","id":"bd0c049702eb4c1687def04167c793b9"},{"url":"http://10.145.65.10/identity_admin","interface":"admin","region":"RegionOne","region_id":"RegionOne","id":"ebca889ec59143abb5f79a066b6f55a7"}],"type":"identity","id":"8c42cb8b0b0f4c9aac185099b0318701","name":"keystone"},{"endpoints":[{"url":"http://10.145.65.10:8774/v2/fdac59f5b20046829ea58720702a74af","interface":"public","region":"RegionOne","region_id":"RegionOne","id":"7a8afadfa68f4e078e055bca8b0da07f"},{"url":"http://10.145.65.10:8774/v2/fdac59f5b20046829ea58720702a74af","interface":"internal","region":"RegionOne","region_id":"RegionOne","id":"a6b035df78b04deba63c8b1684d9ee54"},{"url":"http://10.145.65.10:8774/v2/fdac59f5b20046829ea58720702a74af","interface":"admin","region":"RegionOne","region_id":"RegionOne","id":"efd2dabe1fde4632964f346e35934006"}],"type":"compute_legacy","id":"f2c9f265b1c54f27a94a913cb155f355","name":"nova_legacy"},{"endpoints":[{"url":"http://10.145.65.10:9292","interface":"admin","region":"RegionOne","region_id":"RegionOne","id":"1be1fad26f4e4ffca0bf668a5909f203"},{"url":"http://10.145.65.10:9292","interface":"public","region":"RegionOne","region_id":"RegionOne","id":"5d8c3f7b501147b894b939e5898fa98d"},{"url":"http://10.145.65.10:9292","interface":"internal","region":"RegionOne","region_id":"RegionOne","id":"8b5d0e482e9f44f6aab104c611b59eaf"}],"type":"image","id":"feff310a71b641bf83e069a7885d75c9","name":"glance"}],"tenantId":"fdac59f5b20046829ea58720702a74af"}
*******************
a fake tenantId xxxxxxxxxxx
*******************
2019-04-30 12:06:23,575 INFO [api.call] Response: {"uuid":"5077a800-6afd-11e9-96e3-af1eed1f6296","statusCode":200,"statusMessage":"OK","headers":{"x-powered-by":"Express","access-control-allow-origin":"*","access-control-allow-credentials":"true","content-type":"application/json"},"body":{"monitor":{"id":"fa1f961a-6681-4319-94d8-6478472bc447","name":"t3","description":"My description","tenantId":"fdac59f5b20046829ea58720702a74af","createdAt":"2019-04-28T02:39:28.477Z","updatedAt":null,"interval":5,"timeout":16,"targetAddress":"192.168.10.123","targetPort":8080,"monitorType":"tcp"}}}
```

Therefore the filter does not seem to work at all.

Reply:
I suggest you set a breakpoint at `loopback-datasource-juggler/lib/dao.js:1303` to check why the filter doesn't take effect.

Reply:
@zhang-shengping I write some sample code which can append a new condition to LB4 where clause.

```
  async find(
    @param.query.object('filter', getFilterSchemaFor(Adc)) filter?: Filter,
  ): Promise<CollectionResponse> {
    let f: Filter = filter || {};
    let builder = new WhereBuilder(f.where);
    f.where = builder.and({id: 'abcd'}).build();

    let data = await this.adcRepository.find(f);
    return new CollectionResponse(Adc, data);
  }
```

Reply:
Appending tenantId condition can be regarded as a common operation, which can be implemented in Repository layer. That mean we may input tenantId as an argument when calling Repository function, and Repository function can append tenantId condition to existing filter (or creating a new filter, if no filter input from API).